libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -687,6 +911,37 @@ void AP_UAVCAN::do_cyclic(void)
     	led_out_sem_give();
     }
 
+    if (fix_out_sem_take()) {
+        if (fix_out_array[_uavcan_i] != nullptr &&
+                _broadcast_fix_rate != 0 &&
+                (now - fix_out_send_last_ms) >= fix_out_send_delta_ms) {
+            fix_out_send_last_ms = now;
+            fix_out_send_delta_ms = 1000U / (uint16_t)_broadcast_fix_rate;
+            fix_out_array[_uavcan_i]->broadcast(_fix_state);
+        }
+
+
+        if (fix2_out_array[_uavcan_i] != nullptr &&
+                _broadcast_fix2_rate != 0 &&
+                (now - fix2_out_send_last_ms) >= fix2_out_send_delta_ms) {
+            fix2_out_send_last_ms = now;
+            fix2_out_send_delta_ms = 1000U / (uint16_t)_broadcast_fix2_rate;
+            fix2_out_array[_uavcan_i]->broadcast(_fix2_state);
+        }
+
+        fix_out_sem_give();
+    }
+
+    if (att_out_sem_take() &&
+            attitude_out_array[_uavcan_i] != nullptr &&
+            _broadcast_att_rate != 0 &&
+            (now - att_out_send_last_ms) >= att_out_send_delta_ms) {
+        att_out_send_last_ms = now;
+        att_out_send_delta_ms = 1000U / (uint16_t)_broadcast_att_rate;
+        attitude_out_array[_uavcan_i]->broadcast(_att_state);
+
+        att_out_sem_give();
+	}
```

Comment:
same with do_cyclic. It was gettign really messy with the ESC send stuff so I refactored it so the flow was obvious. Woudl be nice to follow the above examples with a fix_out_send() and att_out_send()

Reply:
Same reasons. It's really straight forward.